File: wear-os/app/src/main/java/com/grantlittman/wearapp/data/model/Signal.kt
package com.grantlittman.wearapp.data.model

/**
 * The type of haptic vibration to produce.
 */
enum class HapticType {
    /** Short, subtle vibration (~50ms). Discreet. */
    TAP,

    /** Medium vibration (~200ms). Noticeable but subtle. */
    BUZZ,

    /** Long, strong vibration (~500ms). Unmistakable. */
    PULSE,

    /** Two quick taps in succession. Easy to distinguish from a single tap. */
    DOUBLE_TAP
}

/**
 * The type of audio tone to play (optional, off by default).
 */
enum class AudioType {
    /** Short, quiet tone. */
    CLICK,

    /** Standard alert tone. */
    BEEP,

    /** Longer, more noticeable tone. */
    CHIME
}

/**
 * A single haptic/audio event the watch can produce.
 *
 * @param hapticType The vibration style.
 * @param intensity Vibration strength from 0.0 (off) to 1.0 (max). Default is 1.0.
 * @param audioType Optional audio tone to accompany the haptic. Null means no audio.
 */
data class Signal(
    val hapticType: HapticType = HapticType.TAP,
    val intensity: Float = 1.0f,
    val audioType: AudioType? = null
)

File: watchos/PulseTimer/PulseTimer Watch App/Models/Signal.swift
import Foundation

/// The type of haptic vibration to produce.
enum HapticType: String, Codable, CaseIterable, Equatable, Hashable {
    case tap = "TAP"
    case buzz = "BUZZ"
    case pulse = "PULSE"
    case doubleTap = "DOUBLE_TAP"
}

/// The type of audio tone to play (optional, off by default).
enum AudioType: String, Codable, CaseIterable, Equatable, Hashable {
    case click = "CLICK"
    case beep = "BEEP"
    case chime = "CHIME"
}

/// A single haptic/audio event the watch can produce.
struct Signal: Codable, Equatable, Hashable {
    var hapticType: HapticType
    /// 0.0 to 1.0. Preserved through serialization for cross-platform
    /// JSON compatibility, but ignored at playback on watchOS: WKInterfaceDevice.play()
    /// does not accept an intensity parameter.
    var intensity: Double
    var audioType: AudioType?

    init(hapticType: HapticType = .tap,
         intensity: Double = 1.0,
         audioType: AudioType? = nil) {
        self.hapticType = hapticType
        self.intensity = intensity
        self.audioType = audioType
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        self.hapticType = (try? c.decode(HapticType.self, forKey: .hapticType)) ?? .tap
        self.intensity = (try? c.decode(Double.self, forKey: .intensity)) ?? 1.0
        self.audioType = try? c.decodeIfPresent(AudioType.self, forKey: .audioType)
    }
}
